src/dune_rpc_impl/server.ml

```diff
@@ -356,6 +358,38 @@ let handler (t : t Fdecl.t) : 'a Dune_rpc_server.Handler.t =
     in
     Handler.request rpc cb Public.Request.diagnostics
   in
+  let () =
+    let cb =
+      let f (`Path path, `Contents contents) =
+        let+ version =
+          Memo.Build.run
```

Comment:
I'm not sure it is safe to call `Memo.Build.run` multiple times in parallel. More precisely, errors might not go to the right place.

Reply:
That's true. Should I use Build_system.run then?

Reply:
Using `Build_system.run` doesn't seem great :/ If there is a build in progress, the request is going to take forever to complete.

I re-read the code of `Memo`, and I think it should actually be ok to `Memo.run`. We should discuss it next week when @snowleopard is back, but in the meantime we can go for it.

We could also have tests for this.

Reply:
If I remember correctly, concurrent calls to `Memo.run` with different error handlers may lead to subtly surprising behaviour, where only one of the handlers receives the errors "early", i.e. all but one handler may receive the errors late, after all concurrent branches have completed.

If `Memo.run` is called at "top-level", i.e. there is no outer `Memo.run_with_error_handler` that sets up a user-specified handler, then Memo uses a trivial error handler that simply ignores the errors.

See a comment next to `val run_with_error_handler` in the `mli`.

Reply:
Ah, we should have just read the doc!

In this case it should be fine then, as this `Memo.run` is called from the RPC which is living outside of the memo world.

Thanks @snowleopard